### PR TITLE
add Classes and Styles props for fields without errors

### PR DESCRIPTION
**Pull request by SimeonGriggs**
...and non-empty values

This adds `validClassName` and `validStyle` to complement the existing `errorClassName` and `errorStyle` to form inputs.

### CODE REVIEW

File: packages/web/src/form/form.js [hunk 1]
```diff
@@ -66,15 +66,22 @@ const inputTagProps = (props) => {
   // any errors on this field
   const validationError = errors[props.name]
 
-  // get errorStyle/errorClassName and replace style/className if present
-  const { errorClassName, errorStyle, ...tagProps } = props
+  // get error/validStyle and error/validClassName and replace style/className if present
+  const { errorClassName, errorStyle, validClassName, validStyle, ...tagProps } = props
```
Comment (jtoar) [inline packages/web/src/form/form.js:70]: This should fix these linting errors:
```suggestion
  const {
    errorClassName,
    errorStyle,
    validClassName,
    validStyle,
    ...tagProps
  } = props
```

File: packages/web/src/form/form.js [hunk 2]
```diff
@@ -66,15 +66,22 @@ const inputTagProps = (props) => {
...
   if (validationError) {
     if (errorClassName) {
       tagProps.className = errorClassName
     }
     if (errorStyle) {
       tagProps.style = errorStyle
     }
+  } else if (!validationError && props.value) {
```
Comment (jtoar) [inline packages/web/src/form/form.js:78]: I see what you're getting at here. I think `props.value` might be the wrong thing to check for (just can't get the validation classes to show up in my app yet&mdash;let me know if I'm doing something wrong though / when I'm supposed to see them). I'm not sure what the right thing is yet; will dig in this weekend.

**Comment by jtoar** [comment]
Glad to see you took this up! :rocket: I'll tag @cannikin just to let him know you're working on this great new feature.

I just pulled this down and tried it out, or tried to try it out at least. I'm still super new to forms, so let me know if I got something wrong while testing it in a Redwood app.

I commented on the exact line, but it looks like the check for `props.value` is blocking `validClass...` name / style from getting through right? I tried removing it and I see why we need it there now&mdash;the valid class just applies right away!

**Comment by SimeonGriggs** [comment]
> I commented on the exact line, but it looks like the check for `props.value` is blocking `validClass...` name / style from getting through right? I tried removing it and I see why we need it there now—the valid class just applies right away! Hence the "...and non-empty values" in your first comment. (I get it now.)

Yeah. So a field without an error might also have no value. But that doesn't mean it's valid. But also, errors are only rendered once `onSubmit` is attempted, and then on each element after an `onBlur` event.

So you can get right down the rabbit hole here on whether you/RJS want error/validation checks on every keystroke event – but this is super annoying for `url` and `email` field types especially. 

I think "this field has a value" && "this field does not contain errors" seems like a good starting point for applying a `valid` class?

**Comment by cannikin** [comment]
Hmm that's tough...I'm trying to think of any examples where a blank value might actually be valid and should be marked as so...but I guess that's not a new problem, the same conundrum would come up for any validator that adds valid classes to form fields.

If you're using `onBlur` to trigger validation, it's going to go through the entire form on every blur, I'm pretty sure. So if blank values are considered valid, those would all get marked as soon as you tabbed out of the first field. That's probably not an ideal experience. But if blanks don't trigger validation (but them being blank is valid and the form could be submitted) then when you tab by those fields, they will NOT get marked as valid, even though they are.

Hmmmmmmmm

**Comment by SimeonGriggs** [comment]
Yeah agreed. So long as it only renders valid classes on field with valid values, I think it's a good idea?

**Comment by cannikin** [comment]
Yeah I think we can give it a shot! 

You should check the functionality of `!validationError && props.value` with a checkbox field. I'm not sure if the value returned is always a string or an actual JS boolean. If it's a string then `"false"` will be `true` as far as that check is concerned and the field will be valid, but if if it's actually a boolean `false` then the field being checked off would NOT be valid.

**Comment by SimeonGriggs** [comment]
> I'm not sure if the value returned is always a string or an actual JS boolean.

Looking at it again today and yeah this is a real problem. Mostly because if the form field values aren't state-managed then there's no `value` in the prop object anyway. 

So my original idea here doesn't seem like it'll work by just piggy-backing off the error classes unfortunately. It's perhaps more work than it initially seemed...

**Comment by cannikin** [comment]
Roger that @SimeonGriggs ... should we close for now?

**Comment by SimeonGriggs** [comment]
Yes I think so...

**Pull request closed without merging**